### DISCUSSION TITLE
How do I used Pandoc to replicate Quarto document input parsing?

### DISCUSSION THREAD

**Discussion by matthew-brett** [quarto-dev/quarto-cli#11393] [Q&A] [answered]
### Description

I'm writing some tests for my project implementing Pandoc filters for Quarto.

I've run into a problem in that I can't see how to replicate the Pandoc AST that Quarto generates, to simulate the input to my filters.

I'm looking for some set of Pandoc input format specifiers that get as close as possible to those that Quarto is using in parsing Markdown documents such as `.qmd` files. 

As an example, I have a document like this:

````qmd
````

Using a null-filter, run `pre-ast` in Quarto, I see this generates the following AST:

```
Div(; classes=['custom-class']) Div(CodeBlock(a = 1; classes=['r', 'cell-code']); classes=['cell'])
```

However, if I just use Pandoc `markdown` as the specified input format, I get this as the AST:

```
Div(; classes=['custom-class']), Para(Code({r} a = 1))
```

Notice that the backtick-code-block has not been correctly parsed to a CodeBlock.  Is there some Pandoc extension or similar I can enable to get the same parsing for this simple document from Pandoc, as I do from Quarto?

**Comment by cscheid** [comment, marked as answer]
> Is there some Pandoc extension or similar I can enable to get the same parsing for this simple document from Pandoc, as I do from Quarto?

Not that we're aware of. Unfortunately, this was a change from Pandoc 2 to 3 which we didn't control or advocate for. In Quarto, we use a [tremendously ugly hack](https://github.com/quarto-dev/quarto-cli/tree/main/src/resources/pandoc/datadir/readqmd.lua) in a custom reader.

One change I want to make in Quarto is to provide a standalone front-end reader that understands our syntax, is used internally, and is usable outside of Quarto itself. This would solve your problem, but it's a big project. I hope to be able to do it in 2025.

**Reply by matthew-brett** [reply]
Thanks - that's helpful.  I'm working around it with my own regex hack for now.

**Reply by matthew-brett** [reply]
The regex hack I used, for reference : https://github.com/matthew-brett/noteout/blob/main/src/noteout/nutils.py#L116
